### PR TITLE
Add support for Super Zero Protocol (SERO) addresses for encoding and decoding

### LINKED ISSUE #193: Add support for Super Zero Protocol (SERO) addresses in encoding library
Here is the encoding library: https://github.com/ensdomains/address-encoder

Example: https://github.com/ensdomains/address-encoder/pull/47/files

Discord for dev help: https://discord.com/invite/AskZbFx

Please provide a reference link so we can check you did the encoding correctly.

We'd like to keep the encoding library as small as possible, so please be efficient in your use of space.

We will likely be reviewing PRs and paying out bounties more or less on a weekly basis. No need to ping us to check your PR, we will definitely get to it.

Thanks for your help!

### PR DESCRIPTION
**Pull request by janus**
## Issue number
closes #193 

## Description
Add two function to handle bs58Encode and bs58Decode. These are no checked.

## List of features added/changed
Checked that the byte length is 96 before accepting decoded value. And it should only be limited to alphabets and numbers.  Length 96 and 64 are the new lengths for SERO address
Reference: https://github.com/sero-cash/go-sero/blob/master/common/address/account_types.go#L81

## How Has This Been Tested?
Add SERO in the test pool so that decoded address should match it text version. Covers length 96

## Checklist:

- [X ] My code follows the code style of this project.
- [ X] My code implements all the required features.

### CODE REVIEW

**Review by makoto — CHANGES_REQUESTED** [review]
`Checked that the byte length is 92 or 64 before accepting decoder value. And it should only be limited to alphabets and numbers. Length 92 and 64 are the new lengths for SERO address` Can you add the reference link to the specification?

**Comment by janus** [comment]
Reference: https://github.com/sero-cash/go-sero/blob/master/common/address/account_types.go#L81

Added.

File: src/index.ts [hunk 1]
```diff
@@ -660,9 +676,11 @@ export const formats: IFormat[] = [
   bitcoinBase58Chain('DIVI', 301, [[0x1e]], [[0xd]]),
   getConfig('DOT', 354, dotAddrEncoder, ksmAddrDecoder),
   getConfig('KSM', 434, ksmAddrEncoder, ksmAddrDecoder),
+  getConfig('SOL', 501, bs58Encode, bs58Decode),
```
Comment (makoto) [inline src/index.ts:679]: I tried to resolve conflict and put duplicate `SOL`. Can you get rid of line 679, keeping line 681?

Comment (makoto) [reply src/index.ts:679]: Umm, now the test is failing. Can you figure out why it is failing?

**Comment by makoto** [comment]
Any update on this?

**Review by makoto — CHANGES_REQUESTED** [review]
Test failing

**Comment by janus** [comment]
Updated. Please check and close.

**Comment by janus** [comment]
> Test failing

Resolved.

File: src/index.ts (hunk 1 above)
Comment (janus) [reply src/index.ts]: Resolved.

File: src/__tests__/index.test.ts [hunk 2]
```diff
@@ -523,6 +523,7 @@ const vectors: Array<TestVector> = [
       { text: 'TUrMmF9Gd4rzrXsQ34ui3Wou94E7HFuJQh', hex: '41cf1ecacaf90a04bb0297f9991ae1262d0a3399e13d6d96c2' }
     ],
   },
+
```
Comment (makoto) [inline src/__tests__/index.test.ts:526]: Please remove extra line

Comment (janus) [reply src/__tests__/index.test.ts]: Rsolved.

File: src/index.ts [hunk 3]
```diff
@@ -465,6 +465,22 @@ function liskAddressDecoder(data: string): Buffer {
...
 }
+  
+function seroAddressEncoder(data: Buffer): string {
+  const address =  bs58EncodeNoCheck(data);
+
+  return address;
+}
+
+function seroAddressDecoder(data: string): Buffer {
+  if (/^[123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]+$/i.test(data)) {
```
Comment (makoto) [inline src/index.ts:476]: Are you doing if the string is bs58 or not? Does `bs58Encode/Decode` do that already?

**Comment by makoto** [comment]
Also https://explorer.sero.cash/accountInfo.html?hash=Tfo8c4aMFbfYhyZgs4hmfHKFVabnWqenhrfmHFr4y2cz8RrQhzkT3YpAiMAnT89wKiiNfUGoY7YCFSZT5PuuQPgNBHJz7WiECRr2n6B6e1a1BfeJ56aPEx2F5aTzwWgBcgx has the length of 131, not 64 nor 92

**Comment by janus** [comment]
> Also https://explorer.sero.cash/accountInfo.html?hash=Tfo8c4aMFbfYhyZgs4hmfHKFVabnWqenhrfmHFr4y2cz8RrQhzkT3YpAiMAnT89wKiiNfUGoY7YCFSZT5PuuQPgNBHJz7WiECRr2n6B6e1a1BfeJ56aPEx2F5aTzwWgBcgx has the length of 131, not 64 nor 92

The invariant holds. I am using 131 length now and my code is still passing. I must have goofed by stating 94 when it is 96 which is what is in my code. The reference  is the master code base for SERO.

File: src/index.ts (hunk 3 above)
Comment (janus) [reply src/index.ts]: Are you doing if the string is bs58 or not? Yes, I am using bs58. Thanks I have cleaned up my code.

**Comment by janus** [comment]
@makoto  Resolved. Go through and merge.

**Comment by makoto** [comment]
I am a bit confused about their go implementation as their test code has completely different address type https://github.com/sero-cash/go-sero/blob/master/zero/account/address_test.go#L26 . I just pinged SERO devs to review this PR so let's wait for a bit before merging this in.

**Comment by janus** [comment]
Okay. Head or tail I would go and rest. I have toiled so much and I have 5 of my entries cancelled out because of code not being available,https://github.com/satoshilabs/slips/blob/master/slip-0044.md and one hanging because of size after multiple reviews. 

Please next time state the caveat so that people like me would assess the risk properly before putting time and energy to work.

 Doesn't look like a good deal at all.

**Comment by makoto** [comment]
Thanks for your effort @janus . I hear your pain. When we started we didn't know what the pain points were so we (at least I) have been learning as we review. It was a bit of mistake to dispatch 100s of issues without knowing what the caveats are. Some of the learning I reflected into the new PR template but we will take your advice on board.

File: src/index.ts [hunk 4]
```diff
@@ -466,6 +466,21 @@ function liskAddressDecoder(data: string): Buffer {
...
+  
+function seroAddressEncoder(data: Buffer): string {
+  const address =  bs58EncodeNoCheck(data);
+
+  return address;
+}
+
+function seroAddressDecoder(data: string): Buffer {
+  const bytes = bs58DecodeNoCheck(data);
+  if (bytes.length === 96) {
```
Comment (makoto) [inline src/index.ts:478]: Why did you get rid of 64? As yo referenced, it does check the length of bytes of 96 and 64 post decoding. https://github.com/sero-cash/go-sero/blob/master/common/address/account_types.go#L95

Comment (janus) [reply src/index.ts:478]: I was like thinking that it may be what made you to think that my code was wrong.

Comment (makoto) [reply src/index.ts:478]: I asked Sero team for the actual specification page and got this one https://wiki.sero.cash/en/index.html?file=Tutorial/sero-exchange-interface#Account

According to the specification, they have `seed(seed) -> sk(Private Key) -> TK(Tracking Key) -> PK(Public Key) -> PKr(Receipt Code)` where the length of the seed is 32,  private key and tk are 64, and public key is 64, and the Pkr is 96.

Comment (janus) [reply src/index.ts:478]: @makoto  Thanks for the update. So should I return to 64 and 96 or just 64?

Comment (makoto) [reply src/index.ts:478]: Just confirmed the team that `Pkr` is what we need to store so your current 96 should be fine. I will merge this now. Thank you for your patience.

**Pull request merged**